src/custom/conversions.rs

```diff
@@ -0,0 +1,246 @@
+use super::{CodegenExtension, CodegenExtsMap};
+
+use anyhow::{anyhow, Result};
+
+use hugr::{
+    extension::{
+        prelude::{sum_with_error, ConstError},
+        simple_op::MakeExtensionOp,
+        ExtensionId,
+    },
+    ops::{constant::Value, custom::ExtensionOp},
+    std_extensions::arithmetic::{
+        conversions::{self, ConvertOpDef},
+        int_types::INT_TYPES,
+    },
+    types::CustomType,
+    HugrView,
+};
+
+use inkwell::types::BasicTypeEnum;
+use inkwell::values::{AnyValue, BasicValue};
+
+use crate::{
+    emit::{
+        func::EmitFuncContext,
+        ops::{emit_custom_unary_op, emit_value},
+        EmitOp, EmitOpArgs,
+    },
+    sum::LLVMSumType,
+    types::TypingSession,
+};
+
+struct ConversionsEmitter<'c, 'd, H>(&'d mut EmitFuncContext<'c, H>);
+
+fn log2(m: u32) -> Option<u8> {
+    if m == 1 {
+        Some(0)
+    } else if m % 2 == 0 {
+        let n = log2(m / 2)?;
+        Some(n + 1)
+    } else {
+        None
+    }
+}
+
+impl<'c, H: HugrView> EmitOp<'c, ExtensionOp, H> for ConversionsEmitter<'c, '_, H> {
+    fn emit(&mut self, args: EmitOpArgs<'c, ExtensionOp, H>) -> Result<()> {
+        let conversion_op = ConvertOpDef::from_optype(&args.node().generalise()).ok_or(anyhow!(
+            "ConversionsEmitter from_optype failed: {:?}",
+            args.node().as_ref()
+        ))?;
+
+        match conversion_op {
+            ConvertOpDef::trunc_u | ConvertOpDef::trunc_s => {
+                let out_sum_ty = args
+                    .outputs
+                    .get_types()
+                    .last()
+                    .map(|a| a.into_struct_type())
+                    .unwrap();
+                let int_ty = out_sum_ty
+                    .get_field_type_at_index(2)
+                    .unwrap()
+                    .into_struct_type()
+                    .get_field_type_at_index(0)
+                    .unwrap()
+                    .into_int_type();
+                let width = int_ty.get_bit_width();
+                let log_width = log2(width).unwrap();
+                let hugr_sum_ty = sum_with_error(vec![INT_TYPES[log_width as usize].clone()]);
+                let sum_ty = LLVMSumType::try_new(&self.0.typing_session(), hugr_sum_ty)?;
+                emit_custom_unary_op(self.0, args, |ctx, arg, _| {
+                    let trunc_result = if conversion_op == ConvertOpDef::trunc_u {
+                        ctx.builder().build_float_to_unsigned_int(
+                            arg.into_float_value(),
+                            int_ty,
+                            "",
+                        )
+                    } else {
+                        // Otherwise it's ConvertOpDef::trunc_s
+                        ctx.builder()
+                            .build_float_to_signed_int(arg.into_float_value(), int_ty, "")
+                    }?
+                    .as_basic_value_enum();
+                    let optional_result = if trunc_result.is_poison() {
```

Comment:
Doesn't this needs to be a runtime check? You can't statically know how big the float is

Also see how these lines have no test coverage

Reply:
😱 here I was thinking inkwell did something magic to make `is_poison` work